asllib/Interpreter.ml

```diff
@@ -560,6 +557,69 @@ module Make (B : Backend.S) = struct
         let v = eval_expr env scope e in
         B.bind_ctrl (B.choice v (return true) (return false)) @@ fun b ->
         if b then continue env else fatal_from e @@ Error.AssertionFailed e
+    | S_While (e,body) ->
+        B.bind_ctrl
+          (B.choice (eval_expr env scope e)
+             (return
+                (fun env ->
+                  eval_seq env scope body s))
+             (return continue))
+          (fun f -> f env)
+    | S_Repeat (body,e) ->
+        eval_seq_kont env scope body
+          (fun env scope ->
+            B.bind_ctrl
+              (B.choice (eval_expr env scope e)
+                 (return continue)
+                 (return (fun env -> eval_stmt env scope s)))
+              (fun f -> f env))
+    | S_For (id,e1,dir,e2,s) ->
+        let* v1  = eval_expr env scope e1 and* v2 = eval_expr env scope e2 in
+        let genv,lenv = env in
+        (* Save old value of `id`, if any *)
+        let oldv = IMap.find_opt id lenv in
```

Comment:
> I don't think that `id` can be in scope before the start of the statement, because:
> 
> > For statements introduce a for-loop counter which is in scope for the body of the for statement. For-loop counters are immutable, execution-time storage elements.
> 
> and
> 
> > A local declaration shall not declare an identifier which is already in scope at the point of declaration.
> 
> So IMHO, we could just overwrite anything that was in before.
> 
> Type-checking the for loop should make sure that it is not in scope before its declaration via the for statement.

As it is an error for a binding of `id` to exist before the start of the statement, I guess that our best option is to follow ASL semantics. That is:
1. Trust typing and just create a binding for the loop indice while starting and destroying it at loop exit.
2. Do not trust typing and flag an fatal error when a binding for `id` exists before the start of the statement.

Notice that 1. would imply to silently destroy the pre-existing binding. I'd rather choose 2., for the sake of defensive programming: of typing is wrong, then the root cause of the  bug will be easier to find.

Reply:
> I would maybe argue that even `for` statements could be seen as a sugar for `while`:
> 
> ```ocaml
> for i = 0 to 12 do
>    doSomething(i);
> end
> ```
> 
> could be just a sugar for
> 
> ```ocaml
> var __i: integer {0 .. 12} = 0;
> while __i <= 12 do
>   let i = __i;
>   __i = __i + 1;
>   doSomething(i);
> end
> ```
> 
> Where `__i` is a fresh variable. This could be avoided by considering `i` as immutable at type-checking but at runtime having an assignment to `i`, for example at type-checking we could rewrite the `S_For` into a `S_While`, and then not having this code in `Interpreter`.
> 
> This does not mean that we have to implement it this way, but maybe that we need to have the same semantics as this

I would not do that. In particular the scoping rule of the loop index looks quite special.

Reply:
> > I don't think that `id` can be in scope before the start of the statement, because:
> > > For statements introduce a for-loop counter which is in scope for the body of the for statement. For-loop counters are immutable, execution-time storage elements.
> > 
> > 
> > and
> > > A local declaration shall not declare an identifier which is already in scope at the point of declaration.
> > 
> > 
> > So IMHO, we could just overwrite anything that was in before.
> > Type-checking the for loop should make sure that it is not in scope before its declaration via the for statement.
> 
> As it is an error for a binding of `id` to exist before the start of the statement, I guess that our best option is to follow ASL semantics. That is:
> 
>     1. Trust typing and just create a binding for the loop indice while starting and destroying it at loop exit.
> 
>     2. Do not trust typing and flag an fatal error when a binding for `id` exists before the start of the statement.
> 
> 
> Notice that 1. would imply to silently destroy the pre-existing binding. I'd rather choose 2., for the sake of defensive programming: of typing is wrong, then the root cause of the bug will be easier to find.

I think you did 2. in the latest patch.

Reply:
> > I would maybe argue that even `for` statements could be seen as a sugar for `while`:
> > ```ocaml
> > for i = 0 to 12 do
> >    doSomething(i);
> > end
> > ```
> > 
> > 
> >     
> >       
> >     
> > 
> >       
> >     
> > 
> >     
> >   
> > could be just a sugar for
> > ```ocaml
> > var __i: integer {0 .. 12} = 0;
> > while __i <= 12 do
> >   let i = __i;
> >   __i = __i + 1;
> >   doSomething(i);
> > end
> > ```
> > 
> > 
> >     
> >       
> >     
> > 
> >       
> >     
> > 
> >     
> >   
> > Where `__i` is a fresh variable. This could be avoided by considering `i` as immutable at type-checking but at runtime having an assignment to `i`, for example at type-checking we could rewrite the `S_For` into a `S_While`, and then not having this code in `Interpreter`.
> > This does not mean that we have to implement it this way, but maybe that we need to have the same semantics as this
> 
> I would not do that. In particular the scoping rule of the loop index looks quite special.

I don't understand how this would change the scoping rule of `i`, as it is clearly defined in the body of the `for` loop. But forget it.

Reply:
That sounds resonnable to me.

Then @maranget, are you confident that those two versions of the same code behaves similarly with this PR?